### PR TITLE
[dtensor] Add propagate_tensor_meta function that skips cache if _are_we_tracing

### PR DESCRIPTION
**Pull request by azahed98**
Fixes an issue where the log softmax handler checked the tensor metadata cache without checking for tracing or symints.

Probably best to merge this after #160798, but not strictly blocking.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

### CODE REVIEW

**Review by xmfan — APPROVED** [review]
Add a comment to not use `_propagate_tensor_meta` directly

File: torch/distributed/tensor/_sharding_prop.py [hunk 1]
```diff
@@ -198,8 +199,25 @@ def _propagate_tensor_meta_non_cached(
     def _propagate_tensor_meta(
         self, op_schema: OpSchema
     ) -> Union[None, TensorMeta, Sequence[Optional[TensorMeta]]]:
+        """
+        Cached version of _propagate_tensor_meta_non_cached
+        This is a private API. Use propagate_tensor_meta instead.
+        """
         return self._propagate_tensor_meta_non_cached(op_schema)
 
+    def propagate_tensor_meta(
```
Comment (bdhirsh) [inline torch/distributed/tensor/_sharding_prop.py:208]: hmm @XilunWu is it an intentional decision in DTensor to have the existing `_propagate_tensor_meta` methods be private? @azahed98 if so we should probably keep things that way (maybe just rename `_propagate_tensor_meta` to `_propagate_tensor_meta_cached`)

Comment (azahed98) [reply torch/distributed/tensor/_sharding_prop.py:208]: PR to rename: https://github.com/pytorch/pytorch/pull/161744
